Define attribute requirement levels (#2522)
open-telemetry/opentelemetry-specification: specification/common/attribute-requirement-level.md

# Attribute Requirement Level

**Status**: [Experimental](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Performance considerations](#performance-considerations)

<!-- tocstop -->

</details>

> _This section applies to Resource, Span, Log, and Metric and describes requirement levels for attributes._

Reviewer: minor: consider alphabetical sorting

```suggestion
_This section applies to Log, Metric, Resource and Span and describes requirement levels for attributes._
```

Following attribute requirement levels are specified:

- **Required**. All instrumentations MUST populate the attribute. Semantic convention defining required attribute expects that an absolute majority instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of a required attribute.

> _Note: Consumers of the telemetry may use `Required` attributes to identify specific semantic convention(s) the given telemetry item follows._

Reviewer: What does "identify" mean? (or maybe consider removing the note from this PR, and use a separate PR to address it later)

Reviewer 2: Identify that a particular semantic convention is used at all. E.g. currently you can check for the presence of the required `db.system` attribute to identify all spans that involve database operations.

Reviewer 3: We probably need to make it clear that a set of semantic conventions is typically associated with an entity (e.g. database) and the presence of a "Required" attribute from that set (e.g. `db.system`) is how you can identify that the telemetry is about that entity.
It doesn't have to be in this PR, can be done later.

- **Conditionally Required**. All instrumentations MUST add the attribute when given condition is satisfied and the attribute value can be [efficiently retrieved and populated](#performance-suggestions). Semantic convention of a `Conditionally Required` level of an attribute MUST clarify the condition under which the attribute is expected to be populated.
`http.route` is an example of a conditionally required attribute to be populated when instrumented HTTP framework provides route information for the instrumented request. Some low-level HTTP server implementations do not support routing and corresponding instrumentations can't populate the attribute.

- **Recommended**. Instrumentations SHOULD add the attribute by default if it's readily available and can be [efficiently populated](#performance-suggestions). Instrumentation MAY offer a configuration option to disable recommended attributes.

> - **Optional**. Instrumentations MAY populate the attribute. Instrumentation SHOULD populate the attribute if and only if user configures instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Optional** attributes.

Reviewer 3: I think the very first sentence is unnecessary (unless I misunderstand the intent).
```suggestion
- **Optional**. Instrumentation SHOULD populate the attribute if and only if user configures instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Optional** attributes.
```

- **Opt-in**. Instrumentations SHOULD NOT add the attribute by default. Instrumentation SHOULD populate the attribute if and only if user configures instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Opt-in** attributes.

The requirement level for attribute is defined by semantic conventions depending on attribute availability across instrumented entities, performance, security, and other factors. When defining requirement levels, semantic conventions MUST take into account signal-specific requirements. For example, Metric attributes that may have high cardinality can only be defined with **Opt-in** level.

> Semantic convention that refers to attribute from another (e.g. general) semantic convention defaults to initial requirement level and MAY modify the requirement level in scope of a child convention.

Reviewer 3: This sentence is not clear to me. What does "defaults" mean? What is a "child convention"?

Author: clarified in the doc

> Instrumentation that cannot populate attribute according to defined requirement level due to performance, security, privacy, or other consideration, MUST use **Opt-in** requirement level on it.

Reviewer: _Instrumentation that cannot populate attribute according to defined requirement level due to ..._ - is this essentially making all the Required attributes `SHOULD` instead of `MUST`?

Author: MUST is still more appropriate - it's semantic convention responsibility to not require something that's not available. But I assume there might be edge-case instrumentation that can't populate specific required attributes. E.g. out-of-process instrumentations may lack some important detail. 
If it's the case and it can't populate the required attribute, such instrumentations should not be even considered as following the specific semantic convention defining this required attribute. 

So I believe it's a better choice to make spec more strict rather than relaxing it for all instrumentations to accommodate rare edge cases.

If over time we see that edge case is more common than we thought, the semantic convention would need to switch the requirement level to conditional or recommended. At this point, it'll go through the normal versioning (potentially we'll need to define new transformation rules to the level).

Author: I removed the exception for `Required` attributes. 
Semantic convention authors should take all the conditions into account when declaring an attribute as `Required` and then it becomes a `MUST` for instrumentation authors to follow.

> ## Performance suggestions

Reviewer 3: Do considerations in this section apply only to "Conditional" level? Does it apply to "Required" or "Recommended"? It is currently referenced only from "Conditional" level. If it is applicable only to "Conditional" level, move it to the definition of "Conditional" or make applicability more clear here.

In some cases instrumented entities do not have attribute value readily available and retrieval can be expensive. For example, `net.peer.name` may require DNS lookup if only IP address is available to the instrumentation.

> Instrumentations MUST NOT perform expensive resource-consuming operation to obtain attribute values. Corresponding attributes MUST NOT be populated by default.

Reviewer: I think performance considerations are more like suggestions/recommendations:

```suggestion
Instrumentations SHOULD NOT perform expensive resource-consuming operation to obtain attribute values. Corresponding attributes MUST NOT be populated by default.
```

> Instrumentations that decide not to populate `Conditionally Required` or `Recommended` attributes due to [performance](#performance-suggestions), security, privacy, or other consideration by default, SHOULD use the **Optional** requirement level on them if the attributes are logically applicable.

Reviewer 4: @lmolkova this also needs to be updated correct? for `Conditionally Required` the instrumentation cannot decide to not populate due to other reason than the condition correct?

Author: thanks! I updated it - I believe we need to give an idea of what happens if the condition is *not* met.

Reviewer 4: +1

# Attribute Requirement Level

**Status**: [Experimental](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

- [Performance suggestions](#performance-suggestions)

<!-- tocstop -->

</details>

_This section applies to Log, and Metric, Resource, and Span describes requirement levels for attributes._

Following attribute requirement levels are specified:

> - **Required**. All instrumentations MUST populate the attribute. Semantic convention defining required attribute expects that an absolute majority instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of required attribute.

Reviewer 2: ```suggestion
- **Required**. All instrumentations MUST populate the attribute. Semantic convention defining required attribute expects that an absolute majority instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of a required attribute.
```

> - **Required**. All instrumentations MUST populate the attribute. Semantic convention defining required attribute expects that an absolute majority instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of a required attribute.

Reviewer 3: I am a bit worried by how strongly this is worded. When designing a semantic convention that needs to be implemented in 10 languages and potentially in dozens of libraries It is hard to be so confident that you understand everything about this attribute and you are certain that this is possible to implement. What happens if in some obscure library it turns out to be impossible to get this data? Does it mean that library cannot support this set of semantic conventions at all or they will just omit this particular required attribute?

I wonder if we need to change `MUST` to `SHOULD` here to recognize the fact that the understanding of semantic convention's author is always going to be imperfect and the developer who implements the convention ultimately should be allowed to make the last call.

Author: @tigrannajaryan 
I understand your concern. I'd like to explain my arguments for MUST and explore this path first.

- the specs we have now are implemented in 10 languages and we can be confident marking specific attributes as required
- we need a process that makes semantic convention owned by OTel stable and implementation of it in a few languages/libraries should be a prereq. I don't think we need 10, but a diverse set of 3-5.
- new semantic conventions can start with everything recommended or conditional and gradually make requirements stricter - it should not be breaking
- SHOULD, despite its description, at least in what I see, is not perceived strong enough
-  If a rare edge-case instrumentation that can't provide a bare-minimum level of attributes should not even be considered as following a specific convention
- consumers of telemetry may want to know what convention(s) this item represents (Azume Monitor needs to for sure). Having a MUST required attribute would help with that

Overall I agree that it makes semantic convention authors' life harder, but I believe it's a good thing that helps instrumentations, vendors, and users.

Reviewer: > I wonder if we need to change `MUST` to `SHOULD` here to recognize the fact that the understanding of semantic convention's author is always going to be imperfect and the developer who implements the convention ultimately should be allowed to make the last call.

+1, IMHO nothing is really `MUST` in semantic convention.

Author: There are quite a few things that come with `MUST` in semantic conventions currently and I don't see any active discussions on changing it.

Reviewer 5: > A rare edge-case instrumentation that can't provide a bare-minimum level of attributes should not even be considered as following a specific convention

For me this is the compromise we should be ready to have if/when such case appears. Is this safe enough? 

(And we should definitely instrument representative enough libraries in a few languages to make sure this kind of components are indeed **commonly** available)

Reviewer 2: I think the semantic conventions could be more useful for consumers of telemetry if there are at least some MUST requirements. If some instrumentation cannot follow the MUST requirements, it does indeed not conform to the semantic conventions. That does not mean it's useless, and there is no MUST requirement for following semantic conventions. It just means that the instrumentation may have issues with some telemetry consumers.

Reviewer 4: Closing this per the discussion in the main thread. @tigrannajaryan feel free to open if you feel otherwise.

> - **Conditional**. All instrumentations SHOULD add the attribute when instrumented entity supports corresponding feature and the attribute value can be [efficiently retrieved and populated](#performance-suggestions).

Reviewer 2: Can we specify that for "conditional" attributes it should (must?) be clarified in the semantic  conventions in which cases they apply/are required? I think this is even currently the case for the old "conditional" attributes.

`http.route` is an example of a conditional attribute: is widely supported by HTTP web frameworks, but some low-level HTTP server implementations do not support it.

> _Note: For producers of telemetry `Required` and `Conditional` levels are semantically the same (under the assumption that `Required` attributes are always available). However, consumers may use this distinction to identify conventions or validate telemetry._

Author: > I am not sure what the purpose of this sentence is. It seems like it can confuse more than help.

I clarified it and moved to `Required` level, I wanted to emphasize that required attributes are something backends can rely upon.

- **Recommended**. Instrumentations SHOULD add the attribute by default if it's readily available and can be [efficiently populated](#performance-suggestions). Instrumentation MAY allow explicit configuration to disable recommended attributes.

- **Opt-in**. Instrumentations SHOULD NOT add the attribute by default. Instrumentation SHOULD populate the attribute if and only if user configures instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Opt-in** attributes.

The requirement level for attribute is defined by semantic conventions depending on attribute availability across instrumented entities, performance, security, and other factors. When defining requirement levels, semantic conventions MUST take into account signal-specific requirements. For example, Metric attributes that may have high cardinality can only be defined with **Opt-in** level.

Semantic convention that refers to an attribute from another (e.g. general) semantic convention MAY modify the requirement level within its own scope. Otherwise, requirement level from referred semantic convention applies.

> Instrumentation that cannot populate `Conditional` or `Recommended` attribute due to performance, security, privacy, or other consideration, MUST use **Opt-in** requirement level on it.

Reviewer 2: I think this does not make sense for "conditional". For example, how should http.route be populated by some low-level HTTP server implementation (to reuse your example), even if the user explicitly requests it?

```suggestion
Instrumentations that decide not to populate `Conditional` or `Recommended` attributes due to performance, security, privacy, or other consideration by default, SHOULD use the **Opt-in** requirement level on them if the attributes are logically applicable.
```

Author: Good question, I don't think it makes sense for `http.route` unless instrumentation lets the user manually set a route in some callback (e.g. based on regex). 
I was thinking about something like `net.peer.name` DNS lookup. Instrumentation can do it more or less efficiently if it caches DNS lookup results. Doing it by default is not a great idea, but if a user enables it knowing the risks, it'd improve telemetry quality.

Reviewer 2: It sounds like something that may often work but not always, and also nothing will break if the instrumentation does not implement it at all. So I think a SHOULD-requirement instead of a MUST here would be enough

Author: Make sense, I use SHOULD clause now

> # Attribute Requirement Level

Reviewer 2: ```suggestion
# Attribute Requirement Levels for Semantic Conventions
```

**Status**: [Experimental](../document-status.md)

<details>
<summary>Table of Contents</summary>

<!-- toc -->

> - [Performance suggestions](#performance-suggestions)

Reviewer 2: Editorial: The toc looks strange. Maybe add (a) subheading(s) for the main section?

Author: Good point, it'd also help us to have anchor links to levels. I did it and moved things around. I did not change any text.

<!-- tocstop -->

</details>

> _This section applies to Log, and Metric, Resource, and Span describes requirement levels for attributes._

Reviewer 3: ```suggestion
_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes._
```

> _This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes._

Reviewer 2: ```suggestion
_This section applies to Log, Metric, Resource, and Span, and describes requirement levels for attributes defined in semantic conventions._
```

> Following attribute requirement levels are specified:

Reviewer 3: ```suggestion
The following attribute requirement levels are specified:
```

> - **Required**. All instrumentations MUST populate the attribute. Semantic convention defining required attribute expects that an absolute majority instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of a required attribute.

Reviewer 3: ```suggestion
- **Required**. All instrumentations MUST populate the attribute. Semantic convention defining a Required attribute expects that an absolute majority of instrumentation libraries and applications are able to efficiently retrieve and populate it, can ensure cardinality, security, and other requirements specific to signal defined by the convention. `http.method` is an example of a Required attribute.
```

> _Note: Consumers of telemetry can detect if telemetry item follows specific semantic convention by checking presence of a `Required` attribute defined by such convention. For example, `db.system` attribute on a span can be used as an indication that span follows database semantics._

Reviewer 3: ```suggestion
_Note: Consumers of telemetry can detect if telemetry item follows a specific semantic convention by checking the presence of a `Required` attribute defined by such convention. For example, the presence of `db.system` attribute on a span can be used as an indication that the span follows database semantics._
```

- **Conditionally Required**. All instrumentations MUST add the attribute when given condition is satisfied. Semantic convention of a `Conditionally Required` level of an attribute MUST clarify the condition under which the attribute is expected to be populated.
`http.route` is an example of a conditionally required attribute to be populated when instrumented HTTP framework provides route information for the instrumented request. Some low-level HTTP server implementations do not support routing and corresponding instrumentations can't populate the attribute.

> - **Recommended**. Instrumentations SHOULD add the attribute by default if it's readily available and can be [efficiently populated](#performance-suggestions). Instrumentation MAY offer a configuration option to disable recommended attributes.

Reviewer 3: ```suggestion
- **Recommended**. Instrumentations SHOULD add the attribute by default if it's readily available and can be [efficiently populated](#performance-suggestions). Instrumentations MAY offer a configuration option to disable Recommended attributes.
```

> - **Optional**. Instrumentation SHOULD populate the attribute if and only if user configures instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Optional** attributes.

Reviewer 3: ```suggestion
- **Optional**. Instrumentations SHOULD populate the attribute if and only if the user configures the instrumentation to do so. Instrumentation that doesn't support configuration MUST NOT populate **Optional** attributes.
```

The requirement level for attribute is defined by semantic conventions depending on attribute availability across instrumented entities, performance, security, and other factors. When defining requirement levels, semantic conventions MUST take into account signal-specific requirements. For example, Metric attributes that may have high cardinality can only be defined with **Optional** level.

> Semantic convention that refers to an attribute from another semantic convention MAY modify the requirement level within its own scope. Otherwise, requirement level from referred semantic convention applies.

Reviewer 3: ```suggestion
Semantic convention that refers to an attribute from another semantic convention MAY modify the requirement level within its own scope. Otherwise, requirement level from the referred semantic convention applies.
```

For example, [Database semantic convention](../trace/semantic_conventions/database.md) references `net.transport` attribute defined in [General attributes](../trace/semantic_conventions/span-general.md) with `Conditionally Required` level on it.

> When condition on `Conditionally Required` attribute is not satisfied and there is no requirement to populate attribute, semantic conventions MAY provide special instructions on how to handle it. If no instructions are given and if instrumentation can populate the attribute, instrumentation SHOULD use the **Optional** requirement level on the attribute.

Reviewer 3: ```suggestion
When the condition on `Conditionally Required` attribute is not satisfied and there is no requirement to populate attribute, semantic conventions MAY provide special instructions on how to handle it. If no instructions are given and if instrumentation can populate the attribute, instrumentation SHOULD use the **Optional** requirement level on the attribute.
```

> For example, `net.peer.name` is `Conditionally Required` by [Database convention](../trace/semantic_conventions/database.md) when available. When only `net.peer.ip` is available,  instrumentation can do DNS lookup, cache and populate `net.peer.name` but only if user explicitly enables instrumentation to do so considering performance issues DNS lookup introduces.

Reviewer 3: ```suggestion
For example, `net.peer.name` is `Conditionally Required` by [Database convention](../trace/semantic_conventions/database.md) when available. When only `net.peer.ip` is available,  instrumentation can do a DNS lookup, cache and populate `net.peer.name` but only if user explicitly enables instrumentation to do so considering performance issues the DNS lookup introduces.
```

> For example, `net.peer.name` is `Conditionally Required` by [Database convention](../trace/semantic_conventions/database.md) when available. When only `net.peer.ip` is available,  instrumentation can do a DNS lookup, cache and populate `net.peer.name` but only if user explicitly enables instrumentation to do so considering performance issues the DNS lookup introduces.

Reviewer 2: ```suggestion
For example, `net.peer.name` is `Conditionally Required` by [Database convention](../trace/semantic_conventions/database.md) when available. When only `net.peer.ip` is available,  instrumentation can do a DNS lookup, cache and populate `net.peer.name` but only if user explicitly enables instrumentation to do so, considering performance issues the DNS lookup introduces.
```
